### PR TITLE
Added LearningRateScheduler functionality for Image Classification

### PR DESCRIPTION
**Pull request by harshithapv**
1. Added LearningRateScheduler abstract class which will be implemented for various learning rate scheduling algorithms. 
2.  Has the lastest Tensorflow nuget which allows Gradient Descent Optimizer to use learning rate input as tensor

### CODE REVIEW

File: src/Microsoft.ML.Dnn/GradientDescentOptimizerTensor.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        {
+            _lr_t = learningRateTensor;
+        }
+
+        public override void _prepare()
+        {
+        }
+
+    }
+}
```
Comment (codemzs) [inline src/Microsoft.ML.Dnn/GradientDescentOptimizerTensor.cs]: I have asked Haiping to add this to the TF .NET library and do a patch release.

File: src/Microsoft.ML.Dnn/ImageClassificationTransform.cs [hunk 2]
```diff
@@ -382,12 +385,18 @@ private IDataView GetShuffledData(string path)
             metrics.Train = new TrainMetrics();
             float accuracy = 0;
             float crossentropy = 0;
+            TrainState trainstate = new TrainState();
+            trainstate.BatchSize = options.BatchSize;
+            trainstate.BatchesPerEpoch = (int)(trainingSet.GetRowCount() ?? 0 / options.BatchSize);
```
Comment (codemzs) [inline src/Microsoft.ML.Dnn/ImageClassificationTransform.cs]: >trainstate.BatchesPerEpoch = (int)(trainingSet.GetRowCount() ?? 0 / options.BatchSize); [](start = 12, length = 87)

GetRowCount() can be null even when there are rows, so this line will result in an incorrect number of batches per epoch. If total number of examples/rows isn't available anywhere then you can simple calculate that in CacheFeaturizedImagesToDisk as we are featuring every image by keeping a counter and then passing that counter to this method.

File: src/Microsoft.ML.Dnn/ImageClassificationTransform.cs [hunk 3]
```diff
@@ -407,15 +416,18 @@ private IDataView GetShuffledData(string path)
 
                         labelBatch[batchIndex] = label;
                         batchIndex += 1;
+                        trainstate.CurrentBatchIndex = batchIndex;
                         // Train.
                         if (batchIndex == batchSize)
                         {
+
+                            learningRate = options.LearningRateScheduler.GetLearningRate(trainstate);
```
Comment (codemzs) [inline src/Microsoft.ML.Dnn/ImageClassificationTransform.cs]: >learningRate = options.LearningRateScheduler.GetLearningRate(trainstate); [](start = 28, length = 73)

please only do this if scheduling is being used otherwise use constant learning as it was being used before. Passing LR in each run is just inefficient and does not make sense if LR is going to be constant.

File: src/Microsoft.ML.Dnn/ImageClassificationTransform.cs [hunk 4]
```diff
@@ -658,10 +672,11 @@ private void VariableSummaries(RefVariable var)
                 {
                     _bottleneckInput = tf.placeholder_with_default(
                         bottleneckTensor,
-                        shape: bottleneckTensorDims ,
+                        shape: bottleneckTensorDims,
                         name: "BottleneckInputPlaceholder");
-                }
+                    _learningRateInput = tf.placeholder(tf.float64, null, name: "learningRateInputPlaceholder");
```
Comment (codemzs) [inline src/Microsoft.ML.Dnn/ImageClassificationTransform.cs]: >_learningRateInput = tf.placeholder(tf.float64, null, name: "learningRateInputPlaceholder"); [](start = 20, length = 92)

This should only be added when scheduling is needed,.

File: src/Microsoft.ML.Dnn/ImageClassificationTransform.cs [hunk 5]
```diff
@@ -710,7 +725,7 @@ private void VariableSummaries(RefVariable var)
 
             tf_with(tf.name_scope("train"), delegate
             {
-                var optimizer = tf.train.GradientDescentOptimizer(learningRate);
+                var optimizer = new GradientDescentOptimizerTensor(_learningRateInput);
```
Comment (codemzs) [inline src/Microsoft.ML.Dnn/ImageClassificationTransform.cs]: > var optimizer = new GradientDescentOptimizerTensor(_learningRateInput); [](start = 15, length = 72)

please pass LR tensor when scheduling is needed.

File: src/Microsoft.ML.Dnn/GradientDescentOptimizerTensor.cs (hunk 1 above)
Comment (Oceania2018) [reply src/Microsoft.ML.Dnn/GradientDescentOptimizerTensor.cs]: Related issue: https://github.com/SciSharp/TensorFlow.NET/issues/422

File: src/Microsoft.ML.Dnn/ImageClassificationTransform.cs (hunk 5 above)
Comment (Oceania2018) [reply src/Microsoft.ML.Dnn/ImageClassificationTransform.cs]: Please upgrade to https://www.nuget.org/packages/TensorFlow.NET/0.11.7

Comment (codemzs) [reply src/Microsoft.ML.Dnn/ImageClassificationTransform.cs]: Thanks Haiping for quickly adding this constructor!!

File: src/Microsoft.ML.Dnn/GradientDescentOptimizerTensor.cs (hunk 1 above)
Comment (harshithapv) [reply src/Microsoft.ML.Dnn/GradientDescentOptimizerTensor.cs]: I am aware of this. I am the one who added the issue on Tensorflow.NET. Thanks.

File: src/Microsoft.ML.Dnn/ImageClassificationTransform.cs [hunk 6]
```diff
@@ -674,7 +688,8 @@ private void VariableSummaries(RefVariable var)
                         bottleneckTensor,
                         shape: bottleneckTensorDims,
                         name: "BottleneckInputPlaceholder");
-                    _learningRateInput = tf.placeholder(tf.float64, null, name: "learningRateInputPlaceholder");
+                    if(useLearningRateScheduler)
+                        _learningRateInput = tf.placeholder(tf.float64, null, name: "learningRateInputPlaceholder");
```
Comment (Oceania2018) [inline src/Microsoft.ML.Dnn/ImageClassificationTransform.cs]: I think we should use `tf.float32`.

Comment (harshithapv) [reply src/Microsoft.ML.Dnn/ImageClassificationTransform.cs]: Thanks for the suggestion. I realize we don't need a high precision to store learning rate. Will change that.

File: build/Dependencies.props [hunk 7]
```diff
@@ -16,15 +16,15 @@
     <GoogleProtobufPackageVersion>3.5.1</GoogleProtobufPackageVersion>
     <LightGBMPackageVersion>2.2.3</LightGBMPackageVersion>
     <MicrosoftExtensionsPackageVersion>2.1.0</MicrosoftExtensionsPackageVersion>
-    <MicrosoftMLOnnxRuntimePackageVersion>0.5.1</MicrosoftMLOnnxRuntimePackageVersion>
+    <MicrosoftMLOnnxRuntimePackageVersion>0.4.0</MicrosoftMLOnnxRuntimePackageVersion>
```
Comment (eerhardt) [inline build/Dependencies.props]: Please revert this change.

File: src/Microsoft.ML.Dnn/LearningRateScheduler.cs [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        internal int CurrentBatchIndex;
+        internal int CurrentEpoch;
+        internal int BatchSize;
+        internal int BatchesPerEpoch;
+    }
+
+    /// <summary>
+    /// This interface defines a learning rate scheduler.
+    /// </summary>
+    public interface ILearningRateScheduler
```
Comment (eerhardt) [inline src/Microsoft.ML.Dnn/LearningRateScheduler.cs]: Instead of an interface, let's use an abstract class here. That way we can add more members to it in the future.

File: src/Microsoft.ML.Dnn/LearningRateScheduler.cs [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    {
+        internal float GetLearningRate(TrainState options);
```
Comment (eerhardt) [inline src/Microsoft.ML.Dnn/LearningRateScheduler.cs]: this should not be `internal`.

File: src/Microsoft.ML.Dnn/LearningRateScheduler.cs [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+        internal int CurrentEpoch;
+        internal int BatchSize;
+        internal int BatchesPerEpoch;
+    }
+
+    /// <summary>
+    /// This interface defines a learning rate scheduler.
+    /// </summary>
+    public abstract class LearningRateScheduler
+    {
```
Comment (eerhardt) [inline src/Microsoft.ML.Dnn/LearningRateScheduler.cs]: Make an `internal` constructor on `LearningRateScheduler`, so no one outside of this assembly can derive from it.

File: src/Microsoft.ML.Dnn/LearningRateScheduler.cs [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+        internal abstract float GetLearningRate(TrainState options);
+    };
```
Comment (eerhardt) [inline src/Microsoft.ML.Dnn/LearningRateScheduler.cs]: This `;` is not necessary.

File: src/Microsoft.ML.Dnn/LearningRateScheduler.cs [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+        /// <summary>
+        /// Learning rate is scaled at epoch boundaries provided in LrSchedule to corresponding multiplier in the LrSchedule.
+        /// Format for LrSchedule: {epoch, scaling factor}
+        /// </summary>
+        public readonly float[,] LrSchedule;
+
+        /// <summary>
+        /// Base Learning rate to start off with.
+        /// </summary>
+        public float BaseLearningRate;
```
Comment (eerhardt) [inline src/Microsoft.ML.Dnn/LearningRateScheduler.cs]: `readonly`?

File: src/Microsoft.ML.Dnn/LearningRateScheduler.cs [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+
+        /// <summary>
+        /// Learning rate decay factor.
+        /// </summary>
+        public float DecayRate;
+
+        /// <summary>
+        /// If Staircase is True the learning rate decays at discrete intervals and the decayed learning rate follows a staircase function.
+        /// </summary>
+        public bool Staircase { get; }
```
Comment (eerhardt) [inline src/Microsoft.ML.Dnn/LearningRateScheduler.cs]: Why is this a property, but everything else in this class is a field?

IMO - we should only be exposing properties publicly in APIs.

File: src/Microsoft.ML.Dnn/LearningRateScheduler.cs [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+    /// Implementation adopted from RESNET-CIFAR benchmark test in Tensorflow slim.
+    /// https://github.com/tensorflow/models/blob/b974c3f95a37acedcc3c58566834c78fcae4b214/official/vision/image_classification/resnet_cifar_main.py
+    /// </summary>
+    public sealed class LsrDecay : LearningRateScheduler
+    {
+        /// <summary>
+        /// Learning rate is scaled at epoch boundaries provided in LrSchedule to corresponding multiplier in the LrSchedule.
+        /// Format for LrSchedule: {epoch, scaling factor}
+        /// </summary>
+        public readonly float[,] LrSchedule;
```
Comment (eerhardt) [inline src/Microsoft.ML.Dnn/LearningRateScheduler.cs]: A thing with having public array fields is that people can change the elements of the array. Even though it is `readonly`, that only means that someone can't change this reference to a different array. But the array itself can be changed.

Comment (eerhardt) [reply src/Microsoft.ML.Dnn/LearningRateScheduler.cs]: Also having a multi-dimensional array `[,]` in a public API is a little bit confusing. If you used an `IReadOnlyList` of some new `struct` type, then you can name each argument:

```
public readonly struct LearningDecayItem
{
    public readonly float Epoch;  // actually should this be float or int?
    public readonly float ScalingFactor;
}
```

Then people know which is which easily.

**Comment by codemzs** [comment]
@harshithapv Can you please update the title and description of the PR?

File: src/Microsoft.ML.Dnn/LearningRateScheduler.cs (hunk 14 above)
Comment (harshithapv) [reply src/Microsoft.ML.Dnn/LearningRateScheduler.cs]: What remedy do you suggest? Basically I want it to be set only in the constructor.

File: src/Microsoft.ML.Dnn/LearningRateScheduler.cs (hunk 13 above)
Comment (harshithapv) [reply src/Microsoft.ML.Dnn/LearningRateScheduler.cs]: Thanks changed it. It doesn't have to be a property. I added that by mistake.

File: src/Microsoft.ML.Dnn/LearningRateScheduler.cs (hunk 11 above)
Comment (harshithapv) [reply src/Microsoft.ML.Dnn/LearningRateScheduler.cs]: Removed

File: src/Microsoft.ML.Dnn/LearningRateScheduler.cs (hunk 14 above)
Comment (harshithapv) [reply src/Microsoft.ML.Dnn/LearningRateScheduler.cs]: Thanks will change that.

File: src/Microsoft.ML.Dnn/ImageClassificationTransform.cs [hunk 15]
```diff
@@ -317,11 +329,18 @@ private IDataView GetShuffledData(string path)
                     .Load(new MultiFileSource(path)));
         }
 
+        private int GetNumSamples(string path)
+        {
+            using (var reader = File.OpenText(path))
+                return int.Parse(reader.ReadLine());
```
Comment (codemzs) [inline src/Microsoft.ML.Dnn/ImageClassificationTransform.cs]: >return int.Parse(reader.ReadLine()); [](start = 16, length = 36)

align with line 334

File: src/Microsoft.ML.Dnn/ImageClassificationTransform.cs [hunk 16]
```diff
@@ -759,6 +805,7 @@ private static IRowMapper Create(IHostEnvironment env, ModelLoadContext ctx, Dat
             scoreColumnName = ctx.Reader.ReadString();
             predictedColumnName = ctx.Reader.ReadString();
             learningRate = ctx.Reader.ReadFloat();
+            useLearningRateScheduling = ctx.Reader.ReadBoolByte();
```
Comment (codemzs) [inline src/Microsoft.ML.Dnn/ImageClassificationTransform.cs]: >useLearningRateScheduling [](start = 12, length = 25)

why are we serializing this? It is only needed at training and not inferencing, right?

File: src/Microsoft.ML.Dnn/ImageClassificationTransform.cs [hunk 17]
```diff
@@ -1172,13 +1222,18 @@ public sealed class TrainMetrics
...
+            public float LearningRate { get; set; }
+
             /// <summary>
             /// String representation of the metrics.
             /// </summary>
             public override string ToString()
             {
                 if (DatasetUsed == ImageClassificationMetrics.Dataset.Train)
-                    return $"Phase: Training, Dataset used: {DatasetUsed.ToString(),10}, Batch Processed Count: {BatchProcessedCount,3}, " +
+                    return $"Phase: Training, Dataset used: {DatasetUsed.ToString(),10}, Batch Processed Count: {BatchProcessedCount,3}, learning rate: {LearningRate,10} " +
```
Comment (codemzs) [inline src/Microsoft.ML.Dnn/ImageClassificationTransform.cs]: >learning rate [](start = 137, length = 13)

Learning Rate

File: src/Microsoft.ML.Dnn/LearningRateScheduler.cs [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+    internal class TrainState
+    {
+        internal int CurrentBatchIndex;
+        internal int CurrentEpoch;
+        internal int BatchSize;
+        internal int BatchesPerEpoch;
+    }
+
+    /// <summary>
+    /// This interface defines a learning rate scheduler.
```
Comment (codemzs) [inline src/Microsoft.ML.Dnn/LearningRateScheduler.cs]: >interface [](start = 13, length = 9)

abstract class

File: src/Microsoft.ML.Dnn/ImageClassificationTransform.cs (hunk 16 above)
Comment (harshithapv) [reply src/Microsoft.ML.Dnn/ImageClassificationTransform.cs]: We should not be serializing this. This was discussed with Zeeshan and will be addressed by him in his PR.

---
In reply to: [337825750](https://github.com/dotnet/machinelearning/pull/4340#discussion_r337825750) [](ancestors = 337825750)
